src/crypto/sharing.cpp

```diff
@@ -0,0 +1,168 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#include "sharing.h"
+
+#include "ccf/crypto/entropy.h"
+
+#include <stdexcept>
+
+namespace crypto
+{
+  /* PRIME FIELD
+
+     For simplicity, we use a finite field F[prime] where all operations
+     are defined in plain uint64_t arithmetic, and we reduce after every
+     operation. This is not meant to be efficient. Compared to e.g. GF(2^n), the
+     main drawback is that we need to hash the "raw secret" to obtain a
+     uniformly-distributed secret.
+  */
+
+  typedef uint64_t element;
+
+  const uint64_t prime = (1ul << 31) - 1ul; // a notorious Mersenne prime
+
+  element reduce(uint64_t x)
+  {
+    return (x % prime);
+  }
+
+  element mul(element x, element y)
+  {
+    return ((x * y) % prime);
+  }
+
+  element add(element x, element y)
+  {
+    return ((x + y) % prime);
+  }
+
+  element sub(element x, element y)
+  {
+    return ((prime + x - y)) % prime;
+  }
+
+  // naive algorithm, used only to compute coefficients, not for use on secrets!
+  element exp(element x, size_t n)
+  {
+    element y = 1;
+    while (n > 0)
+    {
+      if (n & 1)
+        y = mul(y, x);
+      x = mul(x, x);
+      n >>= 1;
+    }
+    return y;
+  }
+
+  element inv(element x)
+  {
+    if (x == 0)
+    {
+      throw std::invalid_argument("division by zero");
+    }
+    return exp(x, prime - 2);
+  }
+
+  // This function is specific to prime=2^31-1.
+  // We assume the lower 31 bits are uniformly distributed,
+  // and retry if they are all set to get uniformity in F[prime].
+
+  element sample()
+  {
+    uint64_t res = prime;
+    while (res == prime)
+    {
+      res = crypto::create_entropy()->random64() & prime;
```

Comment:
I think we'll get a big speedup by reusing this entropy source, rather than creating it fresh each time? And I'm more confident that multiple calls to `random64()` on the _same_ entropy will produce different values (internal state has been swizzled), than that adjacently-constructed entropies will differ on their first value (risk that they're constructed from the same source, but getting different internal copies of the state). That's maybe not clear, but what I mean is:

```
auto ent = crypto::create_entropy();
assert(ent->random64() != ent->random64());
// ^ Surely, definitely, this is the whole point of this rng

assert(crypto::create_entropy()->random64() != crypto::create_entropy()->random64());
 // Probably? Maybe? I'd hope so, but relies on the details of create_entropy(), maybe it needs time between calls!
```

Reply:
I am reasonably sure that's fine with rdseed, which we do use for the in-enclave code. With that said, re-using the object seems like a sensible optimisation here.